### PR TITLE
[CARBONDATA-3989]Fix unnecessary segment files creation when segment is neither updated nor deleted.

### PR DESCRIPTION
**Pull request by akashrn5**
### Why is this PR needed?
1. Unnecessary segment file is created for the segment, even though the segment is neither updated nor deleted.
2. Once the segment files are merged, old segment files are still present and will keep on increasing until clean files is fired.
 
 ### What changes were proposed in this PR?
1. Create the new segment files only when the segment is updated or deleted data.
2. once the segment file merge is done, delete the old segment files to avoid many segment files in the metadata directory.
    
 ### Does this PR introduce any user interface change?
 - No

 ### Is any new testcase added?
 - No
 - Yes

### CODE REVIEW

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/4065/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/2326/

File: integration/spark/src/main/scala/org/apache/carbondata/spark/rdd/CarbonDataRDDFactory.scala [hunk 1]
```diff
@@ -725,43 +725,53 @@ object CarbonDataRDDFactory {
...
+    val updateTableStatusFile = CarbonUpdateUtil.getUpdateStatusFileName(updateModel
+      .updatedTimeStamp.toString)
+    val updatedSegments = SegmentUpdateStatusManager.readLoadMetadata(updateTableStatusFile,
+      carbonTable.getTablePath).map(_.getSegmentName).toSet
     val segmentFiles = segmentDetails.asScala.map { seg =>
-      val load =
-        metadataDetails.find(_.getLoadName.equals(seg.getSegmentNo)).get
-      val segmentFile = load.getSegmentFile
-      var segmentFiles: Seq[CarbonFile] = Seq.empty[CarbonFile]
+      // create new segment files and merge for only updated segments
```
Comment (ajantha-bhat) [inline integration/spark/src/main/scala/org/apache/carbondata/spark/rdd/CarbonDataRDDFactory.scala:733]: I thought you will strictly follow to clean up the base code in the area of modification to make it more readable and maintainable!

here `seg`, `file`, `carbonFile` are not good variable names, you can check and refactor if you want

File: integration/spark/src/main/scala/org/apache/carbondata/spark/rdd/CarbonDataRDDFactory.scala [hunk 2]
```diff
@@ -725,43 +725,53 @@ object CarbonDataRDDFactory {
...
 
-      val segmentMetaDataInfo = segmentMetaDataInfoMap.get(seg.getSegmentNo)
-      val file = SegmentFileStore.writeSegmentFile(
-        carbonTable,
-        seg.getSegmentNo,
-        String.valueOf(System.currentTimeMillis()),
-        load.getPath,
-        segmentMetaDataInfo)
+        val segmentMetaDataInfo = segmentMetaDataInfoMap.get(seg.getSegmentNo)
+        val file = SegmentFileStore.writeSegmentFile(
```
Comment (ajantha-bhat) [inline integration/spark/src/main/scala/org/apache/carbondata/spark/rdd/CarbonDataRDDFactory.scala:741]: here writing segment file again to just update the current timestamp ?
you can add comment If possible.

File: integration/spark/src/main/scala/org/apache/carbondata/spark/rdd/CarbonDataRDDFactory.scala (hunk 1 above)
Comment (akashrn5) [reply integration/spark/src/main/scala/org/apache/carbondata/spark/rdd/CarbonDataRDDFactory.scala:733]: changed

File: integration/spark/src/main/scala/org/apache/carbondata/spark/rdd/CarbonDataRDDFactory.scala (hunk 2 above)
Comment (akashrn5) [reply integration/spark/src/main/scala/org/apache/carbondata/spark/rdd/CarbonDataRDDFactory.scala]: here method level comment says for new updates on the same segment we need to write new segment file with new segmentMetaDataInfo and merge with old segment file , i think that comment is pretty clear. It should be ok i think. Please check once.

**Comment by ajantha-bhat** [comment]
LGTM.

will merge once the build passes

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/4097/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/2356/

**Pull request closed without merging**